text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+    timeout: hc.timeout !== undefined ? hc.timeout.toSeconds() : 5,
+  };
+}
+```
+
+Go translation (Run example in the [Go playground](https://play.golang.org/p/_eRLhVmXmp8):
+
+```go
+package ecs
+
+type HealthCheckIface interface{
+   Command() []string
+   Interval() cdk.Duration;
+   Retries() int
+   StartPeriod() cdk.Duration;
+   Timeout() cdk.Duration;
+}
+
+type HealthCheck struct {
+    command []string
+    interval cdk.Duration;
+    retries int
+    startPeriod cdk.Duration;
+    timeout cdk.Duration;
+}
+
+func (h HealthCheck) Command() []string {
+    return h.command
+}
+
+func (h HealthCheck) Retries() int {
+    return h.retries
+}
+
+func (h HealthCheck) Interval() cdk.Duration {
+    return h.interval
+}
+
+func (h HealthCheck) StartPeriod() cdk.Duration {
+    return h.startPeriod
+}
+
+func (h HealthCheck) Timeout() cdk.Duration {
+    return h.timeout
+}
+
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.Command(),
+        interval:    hc.Interval(),
+        retries:     hc.Retries(),
+        startPeriod: hc.StartPeriod(),
+        timeout:     hc.Timeout(),
+    }
+}
+```
+
+#### Case 2: Extending Typescript datatype interfaces
+
+For Typescript datatype interfaces that extend another datatype interface, the corresponding interfaces and implementing structs would be generated in
+go, with the extending interface embedding the extended one, e.g.:
+
+```ts
+export interface BaseServiceOptions {
+  readonly cluster: ICluster;
+  readonly desiredCount?: number;
+  readonly serviceName?: string;
+  readonly maxHealthyPercent?: number;
+  readonly minHealthyPercent?: number;
+  readonly healthCheckGracePeriod?: Duration;
+  readonly cloudMapOptions?: CloudMapOptions;
+  readonly propagateTags?: PropagatedTagSource;
+  readonly enableECSManagedTags?: boolean;
+  readonly deploymentController?: DeploymentController;
+}
+
+export interface BaseServiceProps extends BaseServiceOptions {
+  readonly launchType: LaunchType;
+}
+```
+
+As before, the Go translation would have the getter methods (e.g. readonly cluster) defined in a the generated interface, which correspond to
+non-exported struct members in the struct that implements each getter.
+
+```go
+package ecs
+
+type BaseServiceOptionsIface interface {
+    Cluster()                ICluster
+    DesiredCount()           *int
+    ServiceName()            *string
+    MaxHealthyPercent()      *int
+    MinHealthyPercent()      *int
+    HealthCheckGracePeriod() *Duration
+    CloudMapOptions()        CloudMapOptions
+    PropagateTags()          PropagatedTagSource
+    EnableECSManagedTags()   *bool
+    DeploymentController()   DeploymentController
+}
+
+type BaseServicePropsIface interface {
+    serviceOptions           BaseServiceOptionsIface  // embeddeded interface
+    LaunchType()             LaunchType
+}
```

Comment:
Should the `BaseServiceOptionsIface` field be anonymous?

Reply:
The original idea was that they would be anonymous (the link to the Go playground example has it as such) but there was some discussion about whether the field should be named. There are some style guides that seem to discourage embedding (E.g. the [Uber style guide](https://github.com/uber-go/guide/blob/master/style.md#avoid-embedding-types-in-public-structs)) -- do you have a recommendation?

Reply:
Flattened interface definitions aren't bad either. But if you want `BaseServicePropsIface` to satisfy the  `BaseServiceOptionsIface` interface, you'll need to add `BaseServiceOptionsIface` methods definitions to `BaseServicePropsIface` somehow.

[Go v1.14](https://golang.org/doc/go1.14) included a language change that removes the annoying clashing Interface method definition issue.

e.g.:

```go
type Foo interface {
     Bar
     SomeMethod() string
}

type Bar interface {
      SomeMethod() string
}
```

prior to go 1.14 this would of been a compile error because of the clashing `SomeMethod`. I think the signatures of the methods still must match, and a interface cannot directly define the same method twice.

Reply:
But with that said CDK for Go is going to have a major problem with interfaces with regard to non-breaking changes. Any time the underlying JSII model for an interface adds a new property it will be a breaking change to the Generated Go interface types.

You may be able to make the interfaces define some unexported method so customers are unable to implement the interface it self, but i think you'll probably have problems with cross package dependencies with that strategy. Since struct `mypkg.Foo` can never implement interface `someotherpkg.Bar` because `Bar` has an unexported method.

```go
type mypkg.Foo struct {
      //...
}

type someotherpkg.Bar {
       mySuperSecreteMethod() err
}

// Compile failure because someotherpkg.Bar.mySuperSecreteMethod is unexported
// and cannot be implemented by mypkg.Foo.
var _ someotherpkg.Bar = (*mypkg.Foo)(nil)
```

Reply:
Basically to address this you probable just have to document the expectation with regard to Go CDK types.  All interfaces can experience breaking changes, and provide guidance on how users should "extend" these interface. Most likely this will involve the customer embedded the relevant interface/struct within their type that they are extending.

Its not the best, but not really sure a better way to do this: https://github.com/jasdel/aws-cdk-go/blob/master/examples/extend_cdk_construct/main.go#L46-L87

Reply:
Though in that linked example I did use [private methods in interface definitions](https://github.com/jasdel/aws-cdk-go/blob/master/cdk/construct.go#L5-L25). This forced customers to use embedding in order to ensure their type satisfied the target interface.  All generated CDK types would follow this embedding and private method pattern as well.

Reply:
@jasdel Updated the section on datatype interfaces as per our discussion last week. I believe we said we wouldn't use the private interface definitions approach for datatypes since that would make extending our own constructs much more difficult -- the same will probably hold true with classes, would you agree?